@xen-orchestra/web-core/lib/components/ui/progress-bar/UiProgressBar.vue

```diff
@@ -0,0 +1,114 @@
+<!-- v2 -->
+<template>
+  <div class="ui-progress-bar" :class="accentClass">
+    <div class="progress-bar">
+      <div class="fill" :style="{ width: `${fillWidth}%` }" />
+    </div>
+    <div v-if="shouldShowSteps" class="steps typo-body-regular-small">
+      <span>{{ $n(0, 'percent') }}</span>
+      <span v-for="step in steps" :key="step">{{ $n(step, 'percent') }}</span>
+    </div>
+    <VtsLegendList class="legend">
+      <UiLegend :accent :value="percentage" unit="%">{{ legend }}</UiLegend>
+    </VtsLegendList>
+  </div>
+</template>
+
+<script lang="ts" setup>
+import VtsLegendList from '@core/components/legend-list/VtsLegendList.vue'
+import UiLegend from '@core/components/ui/legend/UiLegend.vue'
+import { toVariants } from '@core/utils/to-variants.util'
+import { useMax } from '@vueuse/math'
+import { computed } from 'vue'
+
+const {
+  value: _value,
+  max = 100,
+  showSteps,
+} = defineProps<{
+  legend: string
+  value: number
+  max?: number
+  showSteps?: boolean
+}>()
+
+const value = useMax(0, () => _value)
+
+const isInvalid = computed(() => !Number.isFinite(value.value / max) || value.value <= 0)
+
+const percentage = computed(() => (isInvalid.value ? 0 : Math.round((value.value / max) * 100)))
+
+const maxPercentage = computed(() => (isInvalid.value ? 100 : Math.ceil(percentage.value / 100) * 100))
+
+const fillWidth = computed(() => Math.max(0, Math.min((percentage.value / maxPercentage.value) * 100, 100)))
+
+const shouldShowSteps = computed(() => showSteps || (value.value > max && percentage.value > 0))
+
+const steps = computed(() => Math.floor(maxPercentage.value / 100))
```

Comment:
I think these changes would make the core more legible, what do you think?

```suggestion
const percentage = computed(() => (max <= 0 ? 0 : (value.value / max) * 100))

const maxPercentage = computed(() => Math.ceil(percentage.value / 100) * 100)

const fillWidth = useClamp(() => (percentage.value / maxPercentage.value) * 100 || 0, 0, 100)

const shouldShowSteps = computed(() => showSteps || percentage.value > 100)

const steps = useMax(1, () => Math.floor(maxPercentage.value / 100))
```